test/udpDnsCacheTransport.js

```diff
@@ -0,0 +1,158 @@
+          cacheDns: true
+        }), 'client');
+
+        const resolvedHostAddress = '1.1.1.1';
+        let dnsLookupCount = 0;
+        dns.lookup = (host, callback) => {
+          dnsLookupCount++;
+          callback(undefined, resolvedHostAddress);
+        };
+
+        statsd.send('test title', {}, (error) => {
+          assert.equal(error, undefined);
+          setTimeout(() => {
+            assert.equal(dnsLookupCount, 1);
+            assert.equal(socketMock.sendCount, 1);
+            assert.equal(socketMock.host, resolvedHostAddress);
+            assert.equal(socketMock.buf, 'test title');
+            done();
+          }, 1000);
+        });
+      });
+    });
+  });
+
+  describe('Sending messages within TTL', () => {
+    it('should lookup dns once', done => {
+      server = createServer(udpServerType, opts => {
+        const socketMock = mockDgramSocket();
+
+        statsd = createHotShotsClient(Object.assign(opts, {
+          cacheDns: true
+        }), 'client');
+
+        const resolvedHostAddress = '1.1.1.1';
+        let dnsLookupCount = 0;
+        dns.lookup = (host, callback) => {
+          callback(undefined, resolvedHostAddress);
+          dnsLookupCount++;
+        };
+
+        statsd.send('message', {}, (error) => {
+          assert.equal(error, undefined);
+        });
+
+        statsd.send('other message', {}, (error) => {
+          assert.equal(error, undefined);
+          setTimeout(() => {
+            assert.equal(dnsLookupCount, 1);
+            assert.equal(socketMock.sendCount, 2);
+            assert.equal(socketMock.host, resolvedHostAddress);
+            done();
+          }, 1000);
+        });
+      });
+    });
+  });
+
+  describe('Sending messages after TTL expired', () => {
+    it('should lookup dns twice', done => {
+      server = createServer(udpServerType, opts => {
+        const socketMock = mockDgramSocket();
+
+        const dnsTtlMilliseconds = 100;
+        statsd = createHotShotsClient(Object.assign(opts, {
+          cacheDns: true,
+          dnsTtlMilliseconds: dnsTtlMilliseconds
+        }), 'client');
+
+        const resolvedHostAddress = '1.1.1.1';
+        let dnsLookupCount = 0;
+        dns.lookup = (host, callback) => {
+          callback(undefined, resolvedHostAddress);
+          dnsLookupCount++;
+        };
+
+        statsd.send('message', {}, (error) => {
+          assert.equal(error, undefined);
+        });
+
+        statsd.send('other message', {}, (error) => {
+          assert.equal(error, undefined);
+        });
+
+        setTimeout(() => {
+          statsd.send('message 1ms after TTL', {}, (error) => {
+            assert.equal(error, undefined);
+            setTimeout(() => {
+              assert.equal(dnsLookupCount, 2);
+              assert.equal(socketMock.sendCount, 3);
+              done();
+            }, 1000);
+          });
+        }, dnsTtlMilliseconds + 1);
+      });
+    });
+  });
+});
```

Comment:
Great to see, thanks

Reply:
You're welcome! Thanks for looking into this!